File: 2022/src/day11/solution.ts
import { readFileSync } from "fs";
import { join, resolve } from "path";

interface Monkey {
  id: number;
  items: number[];
  operation: (old: number) => number;
  divisor: number;
  iftrue: number;
  iffalse: number;
  items_passed: number;
}

const data = readFileSync(join(resolve(), "./input.txt"), "utf8").split("\n\n");

function getMonkeys(data: string[]): Monkey[] {
  return data.map((line: string) => {
    const lines = line.split("\n");
    return {
      id: Number(lines[0].split("Monkey ")[1].split(":")[0]),
      items: lines[1]
        .split("Starting items: ")[1]
        .split(",")
        .map((i) => Number(i)),
      operation: (old) => {
        let op = lines[2].split("Operation: new = ")[1];
        return eval(op);
      },
      divisor: Number(lines[3].split("Test: divisible by ")[1]),
      iftrue: Number(lines[4].split("If true: throw to monkey ")[1]),
      iffalse: Number(lines[5].split("If false: throw to monkey ")[1]),
      items_passed: 0,
    } as Monkey;
  });
}

function getSol1() {
  const monkeys = getMonkeys(data);
  for (let round = 0; round < 20; round++) {
    for (const monkey of monkeys) {
      while (monkey.items.length > 0) {
        let item = monkey.items.shift();

        const multiplied = Math.floor(monkey.operation(item!) / 3);
        const target =
          multiplied % monkey.divisor === 0 ? monkey.iftrue : monkey.iffalse;
        monkeys[target].items.push(multiplied);
        monkey.items_passed++;
      }
    }
  }

  const sol = monkeys
    .map(({ items_passed }) => items_passed)
    .sort((a, b) => b - a);

  return sol[0] * sol[1];
}

function getSol2() {
  const monkeys = getMonkeys(data);
  let denominator = monkeys
    .map(({ divisor }) => divisor)
    .reduce((a, b) => a * b);

  for (let round = 0; round < 10000; round++) {
    for (const monkey of monkeys) {
      while (monkey.items.length > 0) {
        let item = monkey.items.shift();

        const multiplied = monkey.operation(item!) % denominator;
        const target =
          multiplied % monkey.divisor === 0 ? monkey.iftrue : monkey.iffalse;
        monkeys[target].items.push(multiplied);
        monkey.items_passed++;
      }
    }
  }

  const sol = monkeys
    .map(({ items_passed }) => items_passed)
    .sort((a, b) => b - a);

  return sol[0] * sol[1];
}

console.log(`Sol 1 ${getSol1()}`);
console.log(`Sol 2 ${getSol2()}`);

File: 2022/src/day11/solution.rs
use std::collections::HashMap;

use aoc_2022::read_input;

const DAY: i8 = 11;
const ROUNDS: i8 = 20;

fn eval_expression(exp: String) -> u128 {
    let split: Vec<&str> = exp.split(" ").collect();
    if split.len() != 3 {
        panic!("No no");
    }

    let first = split[0].parse::<u128>().unwrap();
    let second = split[2].parse::<u128>().unwrap();

    match split[1] {
        "+" => first + second,
        "*" => first * second,
        "-" => first - second,
        "/" => first / second,
        _ => panic!("WAT"),
    }
}

#[derive(Debug, Clone)]
struct MonkeyData {
    id: i8,
    initial_items: Vec<u128>,
    operation: String,
    divisible_by: i8,
    to_if_true: i8,
    to_if_false: i8,
}

fn parse_monkey(monkey: &str) -> MonkeyData {
    let monkey_lines = monkey.split("\n").collect::<Vec<&str>>();
    let id = monkey_lines[0]
        .split_once(" ")
        .unwrap()
        .1
        .strip_suffix(":")
        .unwrap()
        .parse::<i8>()
        .unwrap();
    let initial_items: Vec<u128> = monkey_lines[1]
        .split_once(": ")
        .unwrap()
        .1
        .split(", ")
        .map(|item| item.parse::<u128>().unwrap())
        .collect();
    let operation = monkey_lines[2].split_once(" = ").unwrap().1.to_string();

    let divisible_by = monkey_lines[3]
        .split_once("Test: divisible by ")
        .unwrap()
        .1
        .parse::<i8>()
        .unwrap();
    let to_if_true = monkey_lines[4]
        .split_once("If true: throw to monkey ")
        .unwrap()
        .1
        .parse::<i8>()
        .unwrap();

    let to_if_false = monkey_lines[5]
        .split_once("If false: throw to monkey ")
        .unwrap()
        .1
        .parse::<i8>()
        .unwrap();
    // let operated = eval_expression(operation);
    // if (operated as i32).rem_euclid(divisible_by.try_into().unwrap()) == 0 {
    //     monkeys.entry(if_true).or_insert(Vec::new()).push(operated);
    //     println!("Monkey {id} passed {operated} to {if_true}");
    // } else {
    //     let floored = (operated as f64 / 3.0).floor() as usize;
    //     monkeys.entry(if_false).or_insert(Vec::new()).push(floored);
    //     println!("Monkey {id} passed {operated} to {if_false}");
    // }

    MonkeyData {
        id,
        initial_items,
        operation,
        divisible_by,
        to_if_true,
        to_if_false,
    }
}

fn main() {
    let input = read_input(DAY);
    let mut monkeys: Vec<MonkeyData> = input.split("\n\n").map(parse_monkey).collect();
    let mut monkey_inspections: HashMap<i8, usize> = HashMap::with_capacity(10);
    let mut monkey_items: Vec<Vec<u128>> =
        monkeys.iter().map(|m| m.initial_items.clone()).collect();

    for round in 0..2 {
        for monkey in monkeys {
            for item in monkey_items[monkey.id as usize] {
                let operation = monkey.operation.clone().replace("old", &item.to_string());
                let operated = eval_expression(operation);
                if (operated as i32).rem_euclid(monkey.divisible_by.try_into().unwrap()) == 0 {
                    monkeys[monkey.to_if_true as usize]
                        .initial_items
                        .push(operated);
                } else {
                    let floored = (operated as f64 / 3.0).floor() as usize;
                    monkeys[monkey.to_if_false as usize]
                        .initial_items
                        .push(floored as u128);
                }
            }
            println!("{:?}", monkeys)
        }
        println!("{round} {:?}", monkey_inspections)
    }
}
